exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java

```diff
@@ -1221,6 +1223,12 @@ public void interrupted(final InterruptedException e) {
      *   to the user
      */
     public void moveToState(final QueryState newState, final Exception ex) {
+      // if the current thread is the foreman thread, throw an exception
+      // otherwise the foreman will be blocked forever on acceptExternalEvents
+      if (myThreadRef == Thread.currentThread()) {
```

Comment:
Let me ask you this: what was the primary cause of deadlock came out of your investigation?